### PR TITLE
Fix Session state bug and improve Query Efficiency in REPL

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR introduces a bug fix and enhancements to FlintREPL's session management and optimizes query execution methods. It addresses a specific issue where marking a session as 'dead' inadvertently triggered the creation of a new, unnecessary session. This behavior resulted in the new session entering a spin-wait state, leading to duplicate jobs.

The improvements include:
- **Introduction of `earlyExitFlag`**: A new flag, `earlyExitFlag`, has been introduced and is set to `true` under two conditions: when a job is excluded or when it is not associated with the current session's job run ID. This flag is evaluated in the shutdown hook to determine whether the session state should be marked as 'dead'. This change effectively prevents the unintended creation of duplicate sessions by the SQL plugin, ensuring resources are utilized more efficiently.
- **Query Method Optimization**: The method for executing queries has been shifted from scrolling to search, eliminating the need for creating unnecessary scroll contexts. This adjustment enhances the performance and efficiency of query operations.
- **Reversion of Previous Commit**: The PR reverts a previous change (https://github.com/opensearch-project/opensearch-spark/commit/be8202480be851f2aba6d13a2edb12a379b0cd58) following the resolution of the related issue in the SQL plugin (https://github.com/opensearch-project/sql/issues/2436), further streamlining the operation and maintenance of the system.

**Testing**:
1. Integration tests were added to cover both REPL and streaming job functionalities, ensuring the robustness of the fixes.
2. Manual testing was conducted to validate the bug fix.

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/246

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by penghuo** [comment]
> Query Method Optimization: The method for executing queries has been shifted from scrolling to search, eliminating the need for creating unnecessary scroll contexts. This adjustment enhances the performance and efficiency of query operations.

Thanks, just create an issue describe same issue. https://github.com/opensearch-project/opensearch-spark/issues/246

**Comment by penghuo** [comment]
> This change effectively prevents the unintended creation of duplicate sessions by the SQL plugin,
could u elaborate more on this? in which case SQL plugin will create duplicate session?

**Comment by kaituo** [comment]
> > This change effectively prevents the unintended creation of duplicate sessions by the SQL plugin,
> > could u elaborate more on this? in which case SQL plugin will create duplicate session?

FlintREPL has a shutdown hook to set the session state to dead. In our SQL setup, marking a session as 'dead' automatically triggers the creation of a new session. However, the newly created session (initiated by the control plane) will enter a spin-wait state, where it inefficiently waits until timeout, leading to unproductive resource consumption and eventual timeout. This PR added the earlyExitFlag flag and set it to true when the job is excluded or the job is not the current session job run id. Then in the shutdown hook, we check the earlyExitFlag before marking the session state as 'dead'. When this flag is true, it indicates that the control plane has already initiated a new session to handle remaining requests for the current session. Thereby we prevent unnecessary duplicate job creation.

**Review by dai-chen — COMMENTED** [review]
I made single line change on `FlintJob` in https://github.com/opensearch-project/opensearch-spark/pull/241 but was unable to add IT. Could you merge and verify if it works as expected? Thanks!

**Review by penghuo — APPROVED** [review]
Thanks for the change!

File: spark-sql-application/src/main/scala/org/apache/spark/sql/util/EnvironmentProvider.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.apache.spark.sql.util
+
+trait EnvironmentProvider {
```
[the remark this answers is not in the document]
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/util/EnvironmentProvider.scala:11]: added JavaDoc

**Review by dai-chen — APPROVED** [review]
Thanks for the changes!

**Pull request merged**